pkg/operator/admission/validation/pullsecret/validator.go

```diff
@@ -0,0 +1,218 @@
+		return "", "", err
+	}
+
+	resp, err := client.httpClient.Do(req)
+	if err != nil {
+		return "", "", err
+	}
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusUnauthorized {
+		return "", "",
+			fmt.Errorf("unexpected status code for https://%s/v2 , got %d but expected %d or %d",
+				url, resp.StatusCode, http.StatusOK, http.StatusUnauthorized)
+	}
+
+	bearerRealm, service, err := extractValuesFromAuthHeader(resp.Header.Get("WWW-Authenticate"))
+
+	log.Printf("authentication URL returned by %s is %s", url, bearerRealm)
+	return bearerRealm, service, err
+}
+
+func extractValuesFromAuthHeader(header string) (string, string, error) {
+	//if this is not good enough we could get the code from podman.
+	//it is however a lot bigger
+	if !strings.HasPrefix(header, "Bearer realm=\"") {
+		return "", "", fmt.Errorf("header is missing data")
+	}
+	bearerRealm := strings.TrimPrefix(header, "Bearer realm=\"")
+	bearerRealm = bearerRealm[:strings.Index(bearerRealm, "\"")]
+	bearerRealm = strings.TrimSuffix(bearerRealm, "\"")
+
+	splitted := strings.Split(header, ",")
+	if len(splitted) < 2 || len(splitted[1]) == 0 {
+		return "", "", fmt.Errorf("header is missing data")
+	}
+
+	service := strings.Split(header, ",")[1]
+	service = strings.TrimSuffix(service, "\"")
+	service = strings.TrimPrefix(service, "service=\"")
+
+	if len(service) == 0 || len(bearerRealm) == 0 {
+		return bearerRealm, service, fmt.Errorf("header is missing data")
+	}
+
+	return bearerRealm, service, nil
+}
+
+func (client ociRegClient) getToken(log *logrus.Entry, bearerRealm, serviceName, user, password string) (string, error) {
+	req, _ := http.NewRequestWithContext(client.ctx, http.MethodGet, bearerRealm, nil)
+	query := req.URL.Query()
+
+	query.Set("account", url.QueryEscape(user))
+	query.Set("service", serviceName)
+	req.URL.RawQuery = query.Encode()
+	req.SetBasicAuth(user, password)
+
+	resp, err := client.httpClient.Do(req)
+	if err != nil {
+		log.Println(err)
+		return "", err
+	}
+	if resp.StatusCode != http.StatusOK {
+		log.Printf("unsuscessfull call to %s", bearerRealm)
+		return "", fmt.Errorf("authentication unsucessful, got status code %d but wanted %d", resp.StatusCode, http.StatusOK)
+	}
+	log.Printf("suscessfull call to %s", bearerRealm)
+	return decodeResponse(resp.Body)
+}
+
+//validateCredentials validates the required credentials from authStruct which
+//in parallel
+func (client ociRegClient) validateCredentials(authsStruct *authsStruct) []error {
+	expectedResults := 0
+
+	results := make(chan error)
+	defer close(results)
+
+	for k := range authsStruct.Auths {
+		// some urls may need to be ignored.
+		if client.required[k] {
+			expectedResults++
+			go client.validateCredential(client.log, *authsStruct, k, results)
+		}
+	}
+
+	errors := make([]error, 0)
+	for i := 0; i < expectedResults; i++ {
+		routineResult := <-results
+		if routineResult != nil {
+			errors = append(errors, routineResult)
+		}
+	}
+	return errors
+}
```

Comment:
`routineResult := <-results` is actually a common pattern. 

`for range` does not prevent any blocking. because it will wait for the channel to be closed it will block and will be a deadlock because we haven't closed the channel at that point. 

I think we need all the errors, at least for logging to see all repositories that are failing, will add that 

Your suggestion with the waitgroup is equivalent to the current one, just with an extra step.

Reply:
"Also, your solution with an errgroup does not run in parallel because you have not set the limit. So it is not all that much simpler in the end." 

I think this statement is incorrect, [example](https://go.dev/play/p/d3o6Z420M0p) of the official [docs](https://pkg.go.dev/golang.org/x/sync/errgroup#example-Group-JustErrors) uses it without the setLimit. In fact, this is the [source code](https://cs.opensource.google/go/x/sync/+/master:errgroup/errgroup.go;l=60). 

comment of the source code -> "// SetLimit limits the number of active goroutines in this group to at most n."

so it's something that isn't needed here as you want all to run concurrently. 

"Your suggestion with the waitgroup is equivalent to the current one, just with an extra step, and that we now need to pass the waitgroup to the routines." 

No, you don't need to pass the waitGroup to the function, you enclose the call to the go routine in an anonymous function so it can access the waitGroup from inside the anonymous function, like [here](https://gobyexample.com/waitgroups) (this example is just to show we don't need to pass the waitGroup, I know this example is a different case).

"for range does not prevent any blocking. in fact it will wait for the channel to be closed it will block and will be a deadlock because we haven't closed the channel at that point."

Obviously using a "for range" with your code will have a deadlock as you close the channel from the receiver part (something that in general is not a good practice), I didn't propose that. I suggested using a waitgroup + "for range" as it is more idiomatic and clear.

If you want all the errors, I would use a waitGroup like in the official [docs](https://pkg.go.dev/sync#example-WaitGroup) (please note that the wait group is accessible from the anonymous function as I mention before) + the errors channel that you are already using, but using a manual counter to wait on go routines is something that should be avoided as can lead to problems and is not idiomatic.

🙂

Reply:
putting the waitgroup in an anonymous function is even worse. and does not bring *any* value to this code. We know how many results we get , we read as many as we send, it's that simple. why do you want to add a useless step ?